patches/server/1039-Add-more-native-map-methods-to-PDC.patch

```diff
@@ -0,0 +1,260 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Xander <xander.dekeijzer@student.hu.nl>
+Co-authored-by: TheRealRyGuy <ryan@ryguy.me>
+Date: Wed, 11 Oct 2023 20:34:13 +0200
+Subject: [PATCH] Add more native map methods to PDC
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
+index 91cb4a28afa5d83e6de10dab834ed63e2eb3b76f..87e9ee54997c5d0dcb6e9548b7dbb7ffa11a2e80 100644
+--- a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
++++ b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
+@@ -38,7 +38,22 @@ public class CraftPersistentDataContainer implements PersistentDataContainer {
+         Preconditions.checkArgument(type != null, "The provided type cannot be null");
+         Preconditions.checkArgument(value != null, "The provided value cannot be null");
+ 
+-        this.customDataTags.put(key.toString(), this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
++        // Paper start
++        put(key.toString(), type, value);
++    }
++    
++    @Override
++    public <T, Z> void put(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type, Z value) {
++        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
++        Preconditions.checkArgument(type != null, "The provided type cannot be null");
++        Preconditions.checkArgument(value != null, "The provided value cannot be null");
++
++        put(key.toString(), type, value);
++    }
++    
++    private <T, Z> void put(String key, PersistentDataType<T, Z> type, Z value) {
++        this.customDataTags.put(key, this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
++        // Paper end
```

Comment:
I don't think you need the "string" methods anymore because you can just call the put(Key,...) methods from the set(NamespacedKey) method. That way none of the precondition checks are duplicated.

Reply:
Can you explain what you mean?
`    @Override
    public <T, Z> void set(NamespacedKey key, PersistentDataType<T, Z> type, Z value) {
        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
        Preconditions.checkArgument(type != null, "The provided type cannot be null");
        Preconditions.checkArgument(value != null, "The provided value cannot be null");

        this.customDataTags.put(key, this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
    }
    
    @Override
    public <T, Z> void put(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type, Z value) {
        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
        Preconditions.checkArgument(type != null, "The provided type cannot be null");
        Preconditions.checkArgument(value != null, "The provided value cannot be null");

        this.customDataTags.put(key, this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
    }`
    Like this or?

Reply:
no, just call `this.put(key, type, value)` from `set`. `NamespacedKey` extends `Key`.

Reply:
nvm, fixed

Reply:
Wait this is wrong, has now calls itself whoops, can I just cast it to Key to make is call the other has or how should I fix that?

Reply:
Well yes, if the method name is the same, you have to cast. But for set/put its different so its not a concern.

Reply:
Okay, fixed 👍